### PR TITLE
Refactor: Rewrite of the Explorer to use ARIA trees

### PR DESCRIPTION
**Pull request by zorkow**
This is the long awaited rewrite of the Explorer to use the ARIA tree paradigm.

* Speech (and Braille) is compute during enrichment and set in the appropriate `aria-label` or `aria-braillelabel`. Consequently each element has four additonal attributes:
    * `data-semantic-speech` contains speech computed by SRE in ssml markup.
    * `data-semantic-braille` contains the braille computed by SRE (this could be in markup as well in the future, e.g., for 2D output).
    * `aria-label` with the speech without ssml markup
    * `aria-braillelabel` with the braille without markup
* Navigation is now handled directly in the `KeyExplorer` that also handles all subtitle and magnification regions.
* Explorer is started with `Enter` or mouse click inside the structure.
* Functionality ported from the old explorer so far:
     * Arrow key navigation
     * Triggering hyperlinks with `Enter`
     * Self-voicing and synchronised highlighting
* Adds a `SpeechUtil` module for some of the functionality related to speech computation.

*Note, this works with the current version of SRE.*
 
#### Known Issues and Next Steps

* Self-voicing of leaf expressions can cause highlighting bugs when we move to quickly between expressions.  Second highlighter (red) works concurrently and is triggered once the voicing stops. Thus the first highlighter (light blue) might fire first and therefore will not unhighlight a node, when it is red.
* Port other speech functionality from SRE. However some of that needs changes in SRE. This  PR works with the **current** version of SRE.
* We do not recompute the semantic structure in SRE after enrichment.

### CODE REVIEW

File: ts/a11y/semantic-enrich.ts [hunk 1]
```diff
@@ -177,14 +193,34 @@ export function EnrichedMathItemMixin<N, T, D, B extends Constructor<AbstractMat
...
+          this.outputData.braille = this.generator.getSpeech(enriched, enriched);
+          this.inputData.enrichedMml = math.math = this.serializeMml(enriched);
           math.display = this.display;
           math.compile(document);
           this.root = math.root;
         } catch (err) {
           document.options.enrichError(document, this, err);
         }
       }
+      setAria(this.root, document.options.sre.locale);
```
Comment (dpvc) [inline ts/a11y/semantic-enrich.ts:223]: I'm wondering if this call wants to be dependent on the options that cause speech to be generated, since right now this function walks the entire expression tree even if speech hasn't been generated.

**Review by dpvc — APPROVED** [review]
This looks good over all, but I'm concerned about one thing.  The `setAria()` function uses `buildSpeech()`, which in turn uses `ssmlParsing()`, which uses `DOMParser`.  That is only available in the browser, so that may cause problems when semantic-enrich is used in node, though I haven't actually tried it out.

It may be helpful to put `setAria()` behind an option  that disables it for use in node, though it might be nice to have it work there as well.  For that, you would need to use the document adaptor to parse and manipulate the XML.  There are methods for parsing an XML string, and you could use `adaptor.kind()` as a substitute for `node.nodeType` (sort of).  So it looks like it would be possible to make the speech generation even in node.  Of course, the explorer will remain browser-dependent, but it would be nice for semantic-enrich to work for node use.

**Comment by zorkow** [comment]
> This looks good over all, but I'm concerned about one thing. The `setAria()` function uses `buildSpeech()`, which in turn uses `ssmlParsing()`, which uses `DOMParser`. That is only available in the browser, so that may cause problems when semantic-enrich is used in node, though I haven't actually tried it out.
> 

I can use the parsing method from SRE, which selects the correct `DomParser` either from the browser or the `xmldom` library.

File: ts/a11y/SpeechUtil.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+      combine[pros] = (sign === '-') ? orig - relative : orig + relative;
+    }
+  }
+  return combine;
+}
+
+/**
+ * Extracts the prosody value from an attribute.
+ */
+const prosodyRegexp = /([\+|-]*)([0-9]+)%/;
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/SpeechUtil.ts:126]: Old Emacs habit where I use explicit or `\|`.

File: ts/a11y/explorer/KeyExplorer.ts [hunk 3]
```diff
@@ -303,28 +366,40 @@ export class SpeechExplorer extends AbstractKeyExplorer<string> {
     // Make that cleaner and remove force as it is not really used!
     let noUpdate = force;
     force = false;
-    super.Update(force);
-    let options = this.speechGenerator.getOptions();
+    if (!this.active && !force) return;
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/explorer/KeyExplorer.ts:390]: This was from before I commented out the use of the `Speech` method (which might come back into play for locale changes).

File: ts/a11y/SpeechUtil.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  for (let pros of ProsodyKeys) {
+    if (element.hasAttribute(pros)) {
+      let [sign, value] = extractProsody(element.getAttribute(pros));
+      if (!sign) {
+        // TODO: Sort out the base value. It is .5 for volume!
+        combine[pros] = (pros === 'volume') ? .5 : 1;
+        continue;
+      }
+      let orig = prosody[pros] as number;
+      orig = orig ? orig : ((pros === 'volume') ? .5 : 1);
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/SpeechUtil.ts:115]: Yes, it can.  Prosody can have positive  and negative values (e.g., pitch can go higher or lower than the baseline). The are usually only explicitly given if they are non-zero, unless, they change relatively, i.e., pitch goes up by 15% and then goes down by 15%. Then it is at 0.

File: ts/a11y/explorer/KeyExplorer.ts [hunk 5]
```diff
@@ -348,38 +423,73 @@ export class SpeechExplorer extends AbstractKeyExplorer<string> {
...
+        return;
+      }
     }
-    if (code === 32 && event.shiftKey || code === 13) {
-      this.Start();
+    let result = null;
+    if (this.active) {
+      result = this.Move(event);
       this.stopEvent(event);
+      if (result) {
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/explorer/KeyExplorer.ts:482]: It can.

File: ts/output/chtml.ts [hunk 6]
```diff
@@ -85,6 +85,7 @@ CommonOutputJax<
       'white-space': 'nowrap'
     },
 
+    'mjx-container[jax="CHTML"] :focus': {'outline': 'solid 3px'},
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/output/chtml.ts:88]: I've added the corresponding CSS element in `ts/output/svg.ts`:

```JavaScript
    'rect[sre-highlighter-added]': {
      stroke: 'black',
      'stroke-width': '40px'
    }
```

File: ts/a11y/semantic-enrich.ts [hunk 7]
```diff
@@ -163,11 +170,20 @@ export function EnrichedMathItemMixin<N, T, D, B extends Constructor<AbstractMat
...
-        if (document.options.sre.speech !== currentSpeech) {
-          currentSpeech = document.options.sre.speech;
+          // TODO: Sort out the loading of the locales better
+        if (document.options.sre.locale !== currentLocale) {
+          currentLocale = document.options.sre.locale;
+          // TODO: Sort out the loading of the locales better
+          mathjax.retryAfter(
+            Sre.setupEngine({locale: document.options.sre.locale}).then(
+              () => {
+                return Sre.sreReady(); }));
```
[the remark this answers is not in the document]
Comment (zorkow) [reply ts/a11y/semantic-enrich.ts:180]: Probably left over from some `console.log`ing

File: ts/a11y/semantic-enrich.ts (hunk 1 above)
Comment (zorkow) [reply ts/a11y/semantic-enrich.ts:223]: Speech is now generated by default during enrichment and therefore it does not really make sense to not set the `aria` elements. However, I have now added an `enableSpeech` option to the `EnrichedMathDocument` that allows us to switch of speech generation entirely.

**Comment by dpvc** [comment]
I'm good with this, other than two things:

1.  I'm wondering if clicking to enter the walker is a good idea, as I think that will be unexpected by most people.  For example, I know there are people who drag select sections of the math to copy and paste it, and this might be confusing to them (or interfere with that ability).

**Comment by dpvc** [comment]
I'm good with this, other than two things:

1.  I'm wondering if clicking to enter the walker is a good idea, as I think that will be unexpected by most people.  For example, I know there are people who drag select sections of the math to copy and paste it, and this might be confusing to them (or interfere with that ability).  

2.  The fact that this branch incorporates the old v4.0-alpha branch, which is not part of the current develop branch, and was not intended to be merged into develop, is a problem.  You are getting a lot of extra commits that shouldn't be be there.  I will look into making a clean branch that cherry-picks the needed commits from this one, while leaving out the v4.0-alpha commits, and see if I can't straighten that out like I did once before on another branch that was based on the alpha branch.  This one will be a bit trickier, though.

**Comment by dpvc** [comment]
I have made a fresh branch `explorer-rewrite-tmp` (with dashes rather than underscores) that is based of the current `develop` branch, and cherry-picked the commits from this PR that were not already in the `develop` branch.  That gets rid of all the v4.0-alpha noise, and makes this a clean branch without a lot of merging in of other branches along the way.  The command `git diff explorer-rewrite-tmp explorer_rewrite_tmp` shows no differences, so I think everything is as it should be.  I recommend we close this PR and open a new one for `explorer-rewrite-tmp` and make any other changes from there.

If there are other branched based off `v4.0-alpha`, they need to get a similar treatment.  These seem to be 
* `refactor/speechgenerator_pool`
* `complete_explorer_functionality`
* `sre_latex_braille_ext`
* `explorer_rewrite_euro`

The last one is just a single commit from 2cb17082, so easily recreated from the new `explorer-rewrite-tmp` without the v4.0-alpha dependency.  The `sre_latex_braille_ext` branch is already merged into `explorer_rewrite_tmp`, and so is incorporated into this new branch, so `sre_latex_braille_ext` probably is no longer needed.  The other two look like they could be readily rebased off of `explorer-rewrite-tmp`.  So I think it would be good to get these branched cleaned up and remove any unneeded ones.

**Comment by zorkow** [comment]
> 
> If there are other branched based off `v4.0-alpha`, they need to get a similar treatment. These seem to be
> 
> * `refactor/speechgenerator_pool`
> * `complete_explorer_functionality`
> * `sre_latex_braille_ext`
> * `explorer_rewrite_euro`
> 
> The last one is just a single commit from [2cb1708](https://github.com/mathjax/MathJax-src/commit/2cb170820f6e694205b0ee647f5db08960c1bd29), so easily recreated from the new `explorer-rewrite-tmp` without the v4.0-alpha dependency. The `sre_latex_braille_ext` branch is already merged into `explorer_rewrite_tmp`, and so is incorporated into this new branch, so `sre_latex_braille_ext` probably is no longer needed. The other two look like they could be readily rebased off of `explorer-rewrite-tmp`. So I think it would be good to get these branched cleaned up and remove any unneeded ones.

Is really the one `refactor/speechgenerator_pool` that subsumes the others and the `sre_latex_braille_ext` is no longer needed. I will cleanup the branches.
